tests/rptest/tests/upgrade_test.py

```diff
@@ -505,3 +506,101 @@ def verify():
+    def test_upgrade(self):
+        """
+        Verify that there is no data loss with topics that have been 'upgraded',
+        i.e. that is topics pre 22.2 that had remote-write enabled before any
+        cloud retention settings were introduced.
+
+        Their respective new cloud retention settings should be 'infinity' and
+        there should be no truncations or compactions performed on the partitions
+        """
+        total_segments = 20
+        total_bytes = total_segments * self.segment_size
+
+        topic = TopicSpec(name='migrating-topic',
+                          partition_count=1,
+                          replication_factor=3,
+                          cleanup_policy=TopicSpec.CLEANUP_DELETE)
+
+        # Create a topic with a retention size lower then the total proposed
+        # number of bytes to be pushed
+        self.rpk.create_topic(topic.name,
+                              partitions=topic.partition_count,
+                              replicas=topic.replication_factor,
+                              config={
+                                  'segment.bytes': self.segment_size,
+                                  'cleanup.policy': topic.cleanup_policy
+                              })
+
+        # Push data to the topic, add an extra 1024 bytes to push the final segment
+        # past its threshold for archival
+        produce_total_bytes(self.redpanda,
+                            topic=topic.name,
+                            partition_index=0,
+                            bytes_to_produce=total_bytes)
+
+        def cloud_log_size() -> int:
+            s3_snapshot = S3Snapshot([topic], self.redpanda.s3_client,
+                                     self.s3_bucket_name, self.logger)
+            cloud_log_size = s3_snapshot.cloud_log_size_for_ntp(topic.name, 0)
+            self.logger.debug(f"Current cloud log size is: {cloud_log_size}")
+            return cloud_log_size
+
+        # Wait for everything to be uploaded to the cloud.
+        wait_until(lambda: cloud_log_size() >= total_bytes,
+                   timeout_sec=10,
+                   backoff_sec=2,
+                   err_msg=f"Segments not uploaded")
+
+        total_cloud_size_before_upgrade = cloud_log_size()
+
+        # Upgrade and restart all of the nodes
+        self.installer.install(self.redpanda.nodes, RedpandaInstaller.HEAD)
+        self.redpanda.rolling_restart_nodes(self.redpanda.nodes,
+                                            start_timeout=120,
+                                            stop_timeout=120)
+
+        self.redpanda.set_cluster_config(
+            {"cloud_storage_housekeeping_interval_ms": 100},
+            expect_restart=True)
+
+        # Wait for 10 cloud_storage_housekeeping intervals, in the event an erroneous truncation or compaction occurs, it would be done within one of the ticks of this interval
+        time.sleep(10 * 100 * 0.001)
```

Comment:
This is sleeping for 1 second?  Seems unlikely to be deterministic

Reply:
I get that, my concern is that 1 second is so short.  If something bad was happening (e.g. a retention when it shouldn't) then you could easily miss it: even if the housekeeping interval is 100ms, it's unlikely that a slow docker test node is going to reliably do 10x retention passes per second

Reply:
That's a valid point, I was concerned with adding too large a sleep because CI already takes 2+ hrs to run. Does a 5s sleep sound reasonable?

Reply:
For cases like this, I've used `wait_until` and used the timeout as the success path, but I can see why not everyone would like that. Anyway, 5 seconds sounds reasonable to me.